File: recruitment-platform/src/main/java/com/recruitment/platform/service/PublicJobService.java
package com.recruitment.platform.service;

import com.recruitment.platform.common.exception.ResourceNotFoundException;
import com.recruitment.platform.config.RedisConfig;
import com.recruitment.platform.dto.job.PublicJobMapper;
import com.recruitment.platform.dto.job.PublicJobResponse;
import com.recruitment.platform.entity.JobStatus;
import com.recruitment.platform.entity.JobType;
import com.recruitment.platform.repository.JobRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PublicJobService {

    private final JobRepository jobRepository;
    private final PublicJobMapper publicJobMapper;

    /**
     * Cached under CACHE_JOB_LISTINGS (5 min TTL, see RedisConfig) - job search
     * is the highest-traffic read in the whole system (every anonymous visitor
     * and every candidate homepage load hits it), so it's the first thing worth
     * taking off the DB.
     */
    @Cacheable(
            cacheNames = RedisConfig.CACHE_JOB_LISTINGS,
            key = "#keyword + '_' + #location + '_' + #jobType + '_' + #status + '_' + #pageable.pageNumber + '_' + #pageable.pageSize"
    )
    public Page<PublicJobResponse> search(String keyword, String location, JobType jobType, JobStatus status, Pageable pageable) {
        // Public browsing should only ever surface OPEN jobs by default -
        // a null status filter still gets pinned to OPEN here, not "any status",
        // so DRAFT/CLOSED postings never leak to anonymous visitors.
        JobStatus effectiveStatus = status != null ? status : JobStatus.OPEN;
        return jobRepository.search(keyword, location, jobType, effectiveStatus, pageable)
                .map(publicJobMapper::toResponse);
    }

    @Cacheable(cacheNames = RedisConfig.CACHE_JOB_DETAIL, key = "#jobId")
    public PublicJobResponse getById(UUID jobId) {
        return jobRepository.findByIdAndDeletedAtIsNull(jobId)
                .map(publicJobMapper::toResponse)
                .orElseThrow(() -> new ResourceNotFoundException("Job not found."));
    }
}

File: recruitment-platform-UI/src/app/services/job/public-job-service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { JobStatus, JobType, PublicJobResponse } from '../../models/job';
import { environment } from '../../../env/environment';
import { ApiResponse, PageResponse } from '../../models/api-response';


export interface JobSearchParams {
  keyword?: string;
  location?: string;
  jobType?: JobType;
  status?: JobStatus;
  page?: number;
  size?: number;
}

@Injectable({ providedIn: 'root' })
export class PublicJobService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/jobs`;

  search(params: JobSearchParams): Observable<ApiResponse<PageResponse<PublicJobResponse>>> {
    let httpParams = new HttpParams();
    if (params.keyword) httpParams = httpParams.set('keyword', params.keyword);
    if (params.location) httpParams = httpParams.set('location', params.location);
    if (params.jobType) httpParams = httpParams.set('jobType', params.jobType);
    if (params.status) httpParams = httpParams.set('status', params.status);
    httpParams = httpParams.set('page', params.page ?? 0).set('size', params.size ?? 20);

    return this.http.get<ApiResponse<PageResponse<PublicJobResponse>>>(this.baseUrl, { params: httpParams });
  }

  getById(jobId: string): Observable<ApiResponse<PublicJobResponse>> {
    return this.http.get<ApiResponse<PublicJobResponse>>(`${this.baseUrl}/${jobId}`);
  }
}